sw/device/silicon_creator/lib/ownership/ownership_key.c

```diff
@@ -9,39 +9,108 @@
 #include "sw/device/silicon_creator/lib/drivers/flash_ctrl.h"
 #include "sw/device/silicon_creator/lib/drivers/keymgr.h"
 #include "sw/device/silicon_creator/lib/drivers/kmac.h"
-#include "sw/device/silicon_creator/lib/ownership/ecdsa.h"
+#include "sw/device/silicon_creator/lib/nonce.h"
+#include "sw/device/silicon_creator/lib/ownership/owner_verify.h"
+
+#include "flash_ctrl_regs.h"
+#include "hw/top_earlgrey/sw/autogen/top_earlgrey.h"
 
 // RAM copy of the owner INFO pages from flash.
 extern owner_block_t owner_page[2];
 
 OT_WEAK const owner_keydata_t *const kNoOwnerRecoveryKey;
 
-hardened_bool_t ownership_key_validate(size_t page, ownership_key_t key,
-                                       const owner_signature_t *signature,
-                                       const void *message, size_t len) {
+static bool is_signature_empty(const owner_signature_t *signature) {
+  for (size_t i = 0; i < ARRAYSIZE(signature->raw); ++i) {
+    if (signature->raw[i] != 0)
+      return false;
+  }
+  return true;
+}
+
+const owner_detached_signature_t *ownership_signature_scan(
+    uintptr_t start, size_t length, uint32_t command, const nonce_t *nonce) {
+  if (length < sizeof(owner_detached_signature_t))
+    return NULL;
+  length -= sizeof(owner_detached_signature_t);
+  uintptr_t end = start + length;
+  while (start < end) {
+    const owner_detached_signature_t *sig =
+        (const owner_detached_signature_t *)start;
+    if (sig->header.tag == kTlvTagDetachedSignature &&
+        sig->header.length == sizeof(owner_detached_signature_t) &&
+        sig->header.version.major == 0 && sig->command == command &&
+        nonce_equal(&sig->nonce, nonce)) {
+      return sig;
+    }
+    start += sizeof(owner_detached_signature_t);
```

Comment:
Is the intent to scan in increments of 4 pages (e.g. 8192bytes), or should we increment by the equivalent of one flash page?

Reply:
Good catch.  The intention is to scan in flash-page sized increments.

Fixed.